collect_app/src/main/java/org/odk/collect/android/http/HttpInterface.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android.http;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.odk.collect.android.utilities.ResponseMessageParser;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+
+public interface HttpInterface {
+
+    /**
+     * Creates a http connection and sets up an input stream.
+     *
+     * @param uri of the stream
+     * @param contentType check the returned Mime Type to ensure it matches. "text/xml" causes a Hash to be calculated
+     * @return HttpInputStreamResult - An object containing the Stream, Hash and Headers
+     * @throws Exception a multitude of Exceptions such as IOException can be thrown
+     */
+    @NonNull
+    HttpInputStreamResult getHttpInputStream(@NonNull URI uri, @Nullable String contentType) throws Exception;
+
+    /**
+     * Performs a HTTP Head request.
+     *
+     * @param uri of which to perform a HTTP head
+     * @param responseHeaders Map which is populated with the HTTP Headers
+     * @return HTTP status code
+     * @throws Exception a multitude of Exceptions such as IOException can be thrown
+     */
+    int httpHeadRequest(@NonNull URI uri, @NonNull Map<String, String> responseHeaders) throws Exception;
+
+    /**
+     * Uploads files to a Server.
+     *
+     * @param fileList List of Files to be uploaded
+     * @param submissionFile The main file to be uploaded (Form file)
+     * @param uri where to send the submissionFile and fileList
+     * @return ResponseMessageParser object that contains the response XML
+     * @throws IOException can be thrown if files do not exist
+     */
+    ResponseMessageParser uploadSubmissionFile(@NonNull List<File> fileList, @NonNull File submissionFile, @NonNull URI uri) throws IOException;
+
+    /**
+     * Clears the Cookie Stores
+     */
+    void clearCookieStore();
```

Comment:
Why would we want users of this API to be concerned about cookies and the credentials store? 

Initially, the HttpClient forced us to work around its global cookie store and credentials provider but now this is no more needed. We can design an API that doesn't force this schema on us. We had issues with the side-effects of how this worked under concurrent scenarios and we solved the problem by explicitly passing the credentials on each request.

Also, there is no guarantee that the new library that will take care of HTTP interaction has the concept of a global cookie store or credentials provider, which would make the transition harder.

Collect is coupled with these methods in `WebCredentialsUtils.setWebCredentialsFromPreferences()`, for example. 
We could add a new nullable `credentials` arg to `getHttpInputStream()`, `httpHeadRequest()` and `uploadSubmissionsFile()`. Also, the "credentials" concept is linked to Collect's domain, not like the cookie store or the credentials provider, which are linked to the HttpClient library's domain.

This would improve the design in two ways:
- By passing the URL and its credentials at the same time, we explicitly link both concepts, instead of relying to implicit mechanisms based on global state, which is harder to understand and test.
- We would limit the API of `HttpInterface` to the three business operations we need to launch from Collect: get documents, make HEAD pre-requests and upload files. There would be no clutter about things that are not related to business operations.
  This would make the transition to a new library easier by reducing the amount of methods that would have to be overriden and public.

Reply:
The suggestion you made makes a lot of sense, it would make the design cleaner.

I wasn't too happy with httpInterface defining credentials and cookiestore methods, it was only done that way to aid the moving of the httpclient code to one place.

I'll work on a new API as you suggested.

Reply:
Thanks for considering my comments :)

It's great the work you're doing in this PR. It's a great example of good refactoring practices and Kent Beck's famous "make the change easy, then make the easy change".

![image](https://user-images.githubusercontent.com/205913/42806204-260e2390-89ae-11e8-97d5-73fc8ff09625.png)

Reply:
Thanks @ggalmazor - That's a great quote!